File: ios/AzooKeyKeyboard/ReportClient.swift
import Foundation

struct WrongConversionReport {
    let suggested: String
    let selected: String
    let selectedIndex: Int
    let reading: String
    let rawInput: String
    let inputStyle: String
    let leftContext: String
    let rightContext: String
    let japaneseLayout: String
    let textContentType: String
    let returnKeyType: String
}

enum ReportClient {
    private static let legacyWrongConversionEndpoint = URL(
        string: "https://docs.google.com/forms/d/e/1FAIpQLSfpYQqbX8u5SgGVfXjNzCPtKAH_5Mp7PCkUiCiUceEaevb8pQ/formResponse"
    )!

    static func submitSharedConversionImprovement(
        endpoint: String,
        report: WrongConversionReport,
        appVersion: String,
        completion: @escaping @Sendable (Bool) -> Void
    ) {
        let trimmedEndpoint = endpoint.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let url = URL(string: trimmedEndpoint),
              url.scheme == "https",
              url.host != nil else {
            completion(false)
            return
        }
        let payload: [String: Any] = [
            "word": report.selected.trimmingCharacters(in: .whitespacesAndNewlines),
            "ruby": report.reading.trimmingCharacters(in: .whitespacesAndNewlines),
            "importance": 3,
            "categories": [String](),
            "note": "IME候補改善: 第\(report.selectedIndex + 1)候補を選択",
            "source": "Keynako IME",
            "app_version": appVersion,
        ]
        guard let payloadData = try? JSONSerialization.data(withJSONObject: payload, options: [.sortedKeys]) else {
            completion(false)
            return
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("Keynako \(appVersion)", forHTTPHeaderField: "User-Agent")
        request.httpBody = payloadData
        URLSession.shared.dataTask(with: request) { data, response, error in
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            completion(
                error == nil &&
                    (200..<300).contains(status) &&
                    responseAcceptsSubmission(data)
            )
        }.resume()
    }

    static func submitLegacyWrongConversion(
        candidate: String,
        ruby: String,
        index: Int,
        appVersion: String,
        learningEnabled: Bool,
        completion: @escaping @Sendable (Bool) -> Void
    ) {
        var components = URLComponents()
        components.queryItems = [
            URLQueryItem(name: "entry.134904003", value: candidate),
            URLQueryItem(name: "entry.869464972", value: ruby),
            URLQueryItem(name: "entry.1459534202", value: String(index)),
            URLQueryItem(name: "entry.571429448", value: appVersion),
            URLQueryItem(name: "entry.524189292", value: learningEnabled ? "有効" : "無効"),
        ]
        var request = URLRequest(url: legacyWrongConversionEndpoint)
        request.httpMethod = "POST"
        request.setValue("no-cors", forHTTPHeaderField: "mode")
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = components.percentEncodedQuery?.data(using: .utf8)
        URLSession.shared.dataTask(with: request) { _, response, error in
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            completion(error == nil && (200..<400).contains(status))
        }.resume()
    }

    private static func responseAcceptsSubmission(_ data: Data?) -> Bool {
        guard let data, !data.isEmpty else { return true }
        guard data.count <= 64 * 1024,
              let decoded = try? JSONSerialization.jsonObject(with: data) else { return false }
        guard let response = decoded as? [String: Any] else { return true }
        return response["ok"] as? Bool != false
    }
}

File: android/app/src/main/kotlin/io/github/StupidGame/azookey_flutter/ReportClient.kt
package io.github.StupidGame.azookey_flutter

import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

data class WrongConversionReport(
    val suggested: String,
    val selected: String,
    val selectedIndex: Int,
    val reading: String,
    val rawInput: String,
    val inputStyle: String,
    val leftContext: String,
    val rightContext: String,
    val japaneseLayout: String,
    val textContentType: String,
    val returnKeyType: String,
)

object ReportClient {
    private const val LEGACY_WRONG_CONVERSION_ENDPOINT =
        "https://docs.google.com/forms/d/e/1FAIpQLSfpYQqbX8u5SgGVfXjNzCPtKAH_5Mp7PCkUiCiUceEaevb8pQ/formResponse"
    private const val MAXIMUM_RESPONSE_BYTES = 64 * 1024

    fun submitSharedConversionImprovement(
        endpoint: String,
        report: WrongConversionReport,
        appVersion: String,
        completion: (Boolean) -> Unit,
    ) {
        Thread {
            completion(
                try {
                    val url = URL(endpoint.trim())
                    require(url.protocol == "https" && url.host.isNotBlank())
                    val body = JSONObject(sharedConversionImprovementPayload(report, appVersion))
                        .toString()
                    val connection = (url.openConnection() as HttpURLConnection).apply {
                        requestMethod = "POST"
                        connectTimeout = 15_000
                        readTimeout = 30_000
                        doOutput = true
                        instanceFollowRedirects = true
                        setRequestProperty("Content-Type", "application/json; charset=utf-8")
                        setRequestProperty("Accept", "application/json")
                        setRequestProperty("User-Agent", "Keynako $appVersion")
                    }
                    try {
                        connection.outputStream.use { it.write(body.toByteArray(StandardCharsets.UTF_8)) }
                        val status = connection.responseCode
                        val stream = if (status >= 400) connection.errorStream else connection.inputStream
                        val bytes = stream?.use { it.readNBytes(MAXIMUM_RESPONSE_BYTES + 1) } ?: byteArrayOf()
                        status in 200..299 &&
                            bytes.size <= MAXIMUM_RESPONSE_BYTES &&
                            responseAcceptsSubmission(String(bytes, StandardCharsets.UTF_8))
                    } finally {
                        connection.disconnect()
                    }
                } catch (_: Exception) {
                    false
                },
            )
        }.start()
    }

    fun submitLegacyWrongConversion(
        candidate: String,
        ruby: String,
        index: Int,
        appVersion: String,
        learningEnabled: Boolean,
        completion: (Boolean) -> Unit,
    ) {
        Thread {
            completion(
                try {
                    val body = formEncode(
                        listOf(
                            "entry.134904003" to candidate,
                            "entry.869464972" to ruby,
                            "entry.1459534202" to index.toString(),
                            "entry.571429448" to appVersion,
                            "entry.524189292" to if (learningEnabled) "有効" else "無効",
                        ),
                    )
                    val connection = (URL(LEGACY_WRONG_CONVERSION_ENDPOINT).openConnection() as HttpURLConnection).apply {
                        requestMethod = "POST"
                        connectTimeout = 15_000
                        readTimeout = 15_000
                        doOutput = true
                        setRequestProperty("mode", "no-cors")
                        setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
                    }
                    connection.outputStream.use { it.write(body.toByteArray(StandardCharsets.UTF_8)) }
                    val status = connection.responseCode
                    connection.disconnect()
                    status in 200..399
                } catch (_: Exception) {
                    false
                },
            )
        }.start()
    }

    internal fun sharedConversionImprovementPayload(
        report: WrongConversionReport,
        appVersion: String,
    ): Map<String, Any> = linkedMapOf(
        "word" to report.selected.trim(),
        "ruby" to report.reading.trim(),
        "importance" to 3,
        "categories" to emptyList<String>(),
        "note" to "IME候補改善: 第${report.selectedIndex + 1}候補を選択",
        "source" to "Keynako IME",
        "app_version" to appVersion,
    )

    private fun responseAcceptsSubmission(body: String): Boolean {
        if (body.isBlank()) return true
        return try {
            val response = JSONObject(body)
            !response.has("ok") || response.optBoolean("ok", true)
        } catch (_: Exception) {
            false
        }
    }

    private fun formEncode(entries: List<Pair<String, String>>): String = entries.joinToString("&") { (key, value) ->
        "${URLEncoder.encode(key, StandardCharsets.UTF_8.name())}=${URLEncoder.encode(value, StandardCharsets.UTF_8.name())}"
    }

}
